designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,499 @@
+
+## Goals
+
+There are two goals with this design:
+
+1. **Make it easier to configure ESLint.** For users, we want to reduce unnecessarily boilerplate and guesswork as much as possible.
+1. **Encourage plugins to use `configs`.** For plugin authors, we want to encourage the use of a consistent entrypoint, `configs`, where users can find predefined configurations. 
+
+## Detailed Design
+
+Design Summary:
+
+1. Allow arrays in config arrays (in addition to objects)
+1. Introduce an `extends` key in config objects
+
+### Allow Arrays in Config Arrays
+
+This was part of the original RFC but was disabled over concerns about the complexity involved with importing configs from packages and having no idea whether it was an object, an array, or an array with a mix of objects and arrays. This mattered most when users needed to use `map()` to apply different `files` or `ignores` to configs, but with `extends`, that concern goes away.
+
+Support for nested arrays is already in `ConfigArray`, it's just disabled.
+
+### Introduce an `extends` Key
+
+The `extends` key is an array that may contain other configurations. When used, ESLint will expand the config object into multiple config objects in order to create a flat structure to evaluate. 
+
+#### Extending Objects
+
+You can pass one or more objects in `extends`, like this:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+Assuming these config objects do not contain `files` or `ignores`, ESLint will convert this config into the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+    {
+        ...js.configs.recommended
+        files: ["**/src/*.js"],
+    },
+    {
+        ...example.configs.recommended
+        files: ["**/src/*.js"],
+    },
+];
+```
+
+If the objects in `extends` contain `files` or `ignores`, then ESLint will merge those values with the values found in the config using `extends`. For example:
+
+```js
+import globals from "globals";
+
+const config1 = {
+    name: "config1",
+    files: ["**/*.cjs"],
+    languageOptions: {
+        sourceType: "commonjs",
+        globals: {
+            ...globals.node
+        }
+    }
+};
+
+const config2 = {
+    name: "config2",
+    files: ["**/*.js"],
+    ignores: ["**/tests"],
+    rules: {
+        "no-console": "error"
+    }
+};
+
+
+export default [
+
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        extends: [config1, config2],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Here, the `files` keys will be combined and the `ignores` key will be inherited, resulting in a final config that looks like this:
```

Comment:
Worth noting that overriding is the behaviour we went with for `typescript-eslint`'s `extends` property and so far we haven't had anyone report issues asking for a merge instead (it has been live for ~9 months now)

Reply:
how to merge:
```js
{ files: ["src/**/*"], extends: [{files: ["**/*.js"]}], }
```

* option A: `files: ["src/**/*"]`
* option B: `files: ["**/*.js"]`
* option C: `files: ["src/**/*", "**/*.js"]`

IMHO, all these are inappropriate. it's expected: `files: ["src/**/*.js"]`

Reply:
IMO Option A is the correct result. It's the least magical and most straightforward answer.

Option C is what this RFC currently proposes.

Reply:
I think a typical use case for shareable configs specifying `files` is to provide different configs for different file extensions, while a typical use case for end-user configs specifying `files` along with `extends` would be to restrict the sharable config to certain directories only, so what would make the most sense, I think, is resulting config that matches the intersection of files, i.e., files that match both user-specified `files` and sharable config's `files`. In @aladdin-add's example, that would be files that match `src/**/*` AND `**/*.js`, which is indeed `files: ["src/**/*.js"]` (though I'm not sure if we would be able to implement this kind of calculations, but rather introduce another mechanism for specifying intersections in the resulting config).

Reply:
Intersection is also how `.eslintrc`'s extends works:

https://stackblitz.com/edit/stackblitz-starters-e4kgkq?file=.eslintrc.json,shareable-config.json

Reply:
I was surprised by typescript-eslint overriding extended `files` patterns. It was not what I wanted or expected.

My 2 cents, reiterating my comment here: https://github.com/eslint/eslint/issues/19116#issuecomment-2473574008

If an extended config specifies `files`, and I also specify `files`, I expect both patterns to apply.

It's easy enough to understand: the extended config filters the list of input files, then my config filters it again. This seems useful.

It's easy enough to implement: instead of glob merging, just use an array-of-arrays format that requires input files to match **any** pattern in **every** array.

It's also easy enough to provide an escape hatch like `files: () => ["src/**/*"]` when you don't want the merging behaviour.

Reply:
We do actually have an intersection syntax already supported in flat config (see https://github.com/eslint/eslint/issues/18966), which we were planning on removing in v10 because we weren't using it. We can certainly leverage that for this proposal as I find the arguments of @mdjermanovic and @aaronadamsCA compelling.

@bradzacher I did take a look at what you were doing with `extends` and I don't believe that's the correct behavior.